lightning/backend/local.go

```diff
@@ -1002,44 +1082,45 @@ loopWrite:
 func (local *local) writeAndIngestByRanges(ctx context.Context, engineFile *LocalFile, ranges []Range, remainRanges *syncdRanges) error {
 	if engineFile.Length == 0 {
 		// engine is empty, this is likes because it's a index engine but the table contains no index
-		log.L().Warn("engine contains no data", zap.Stringer("uuid", engineFile.Uuid))
+		log.L().Info("engine contains no data", zap.Stringer("uuid", engineFile.Uuid))
 		return nil
 	}
 	log.L().Debug("the ranges Length write to tikv", zap.Int("Length", len(ranges)))
 
-	errCh := make(chan error, len(ranges))
+	var allErrLock sync.Mutex
+	var allErr error
+	var wg sync.WaitGroup
+
+	wg.Add(len(ranges))
 
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
 	for _, r := range ranges {
 		startKey := r.start
 		endKey := r.end
 		w := local.rangeConcurrency.Apply()
 		go func(w *worker.Worker) {
-			defer local.rangeConcurrency.Recycle(w)
+			defer func() {
+				local.rangeConcurrency.Recycle(w)
+				wg.Done()
+			}()
 			var err error
 			for i := 0; i < maxRetryTimes; i++ {
-				if err = local.writeAndIngestByRange(ctx, engineFile, startKey, endKey, remainRanges); err != nil {
-					log.L().Warn("write and ingest by range failed",
-						zap.Int("retry time", i+1), log.ShortError(err))
-				} else {
-					break
+				err = local.writeAndIngestByRange(ctx, engineFile, startKey, endKey, remainRanges)
+				if err == nil {
+					return
 				}
+				log.L().Warn("write and ingest by range failed", zap.Int("retry time", i+1), log.ShortError(err))
 			}
-			errCh <- err
+
+			allErrLock.Lock()
+			allErr = multierr.Append(allErr, err)
+			allErrLock.Unlock()
 		}(w)
 	}
 
-	var err error
-	for i := 0; i < len(ranges); i++ {
-		// wait for all sub tasks finish to avoid panic. if we return on the first error,
-		// the outer tasks may close the pebble db but some sub tasks still read from the db
-		e := <-errCh
-		if e != nil && err == nil {
-			err = e
-		}
-	}
-	return err
+	// wait for all sub tasks finish to avoid panic. if we return on the first error,
```

Comment:
could we use another way to support both return early and don't panic, such as recovering, peeking context? I'm not sure if these calls to pd and TiKV is slow

Reply:
dunno. but this part is basically just a refactor of the existing code.